flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientHandlerTest.java

```diff
@@ -208,6 +211,53 @@ public void testCancelBeforeActive() throws Exception {
 		client.cancelRequestFor(inputChannel.getInputChannelId());
 	}
 
+	/**
+	 * Verifies that {@link RemoteInputChannel} is enqueued in the pipeline, and
+	 * {@link AddCredit} message is sent to the producer.
+	 */
+	@Test
+	public void testNotifyCreditAvailable() throws Exception {
+		final CreditBasedClientHandler handler = new CreditBasedClientHandler();
+		final EmbeddedChannel channel = new EmbeddedChannel(handler);
+
+		final RemoteInputChannel inputChannel = mock(RemoteInputChannel.class);
```

Comment:
Could you use `RemoteInputChannel` instead of mocking it? You could add test method `createTestRemoteInputChannel()` and it seems like you would only need to mock `SingleInputGate`

Reply:
I reviewed all the previous existing tests in `PartitionRequestClientHandlerTest` and all the `RemoteInputChannel` is mocked, so I also used the mocked channel here.

I think the mocked input channel is enough for this test, because there is no interaction with internal input channel. The handler just enqueues this input channel and pools it to send `AddCredit` message.

If you prefer the real input channel, I am willing to modify it.

Reply:
I know that mocking was extensively used so far, but we had quite bad experience with it.

I don't agree that it is not interacting, because it is for example checking `inputChannel.isReleased()`, which you need to mock one test below. Now imagine that during some refactor/implementing new feature someone modifies handlers code, by adding some additional interaction/sanity check on `inputChannel`. With mocks it will lead with thousands of changes in all of the mocked unit tests.

Reply:
Thanks for sharing your idea. 
I think I can understand your consideration and agree with that to some extent.

In my current new added tests, I only want to verify the behaviors for released channel and un-released channel in handler. So it is easy to just control `isReleased` property in mocked `InputChannel`, and does not concern the details in `InputChannel`.

As you said, if someone adds more interactions with `InputChannel` in the future or so, the previous tests with mocked `InputChannel` may bring potential risks. Considering this, I am willing to modify the new added tests with real `InputChannel`. :)